drivers/counter/counter_sam0_tc32.c

```diff
@@ -0,0 +1,501 @@
+{
+	struct counter_sam0_tc32_data *data = DEV_DATA(dev);
+	const struct counter_sam0_tc32_config *const cfg = DEV_CFG(dev);
+	TcCount32 *tc = cfg->regs;
+
+	if (chan_id != 0) {
+		return -EINVAL;
+	}
+
+	if (alarm_cfg->ticks > tc->CC[0].reg) {
+		return -EINVAL;
+	}
+
+	int key = irq_lock();
+
+	if (data->ch.callback) {
+		goto err_busy;
+	}
+
+	data->ch.callback = alarm_cfg->callback;
+	data->ch.user_data = alarm_cfg->user_data;
+
+	if (alarm_cfg->absolute) {
+		tc->CC[1].reg = alarm_cfg->ticks;
+		wait_synchronization(tc);
+		tc->INTFLAG.reg = TC_INTFLAG_MC1;
+		tc->INTENSET.reg = TC_INTFLAG_MC1;
+	} else {
+		counter_sam0_tc32_relative_alarm(dev, alarm_cfg->ticks);
+	}
+
+	irq_unlock(key);
+
+	return 0;
+
+err_busy:
+	irq_unlock(key);
+	return -EBUSY;
+}
+
+static int counter_sam0_tc32_cancel_alarm(struct device *dev, u8_t chan_id)
+{
+	struct counter_sam0_tc32_data *data = DEV_DATA(dev);
+	const struct counter_sam0_tc32_config *const cfg = DEV_CFG(dev);
+	TcCount32 *tc = cfg->regs;
+
+	if (chan_id != 0) {
+		return -EINVAL;
+	}
+
+	int key = irq_lock();
+
+	data->ch.callback = NULL;
+	tc->INTENCLR.reg = TC_INTENCLR_MC1;
+	tc->INTFLAG.reg = TC_INTFLAG_MC1;
+
+	irq_unlock(key);
+	return 0;
+}
+
+static int counter_sam0_tc32_set_top_value(struct device *dev, u32_t ticks,
+		counter_top_callback_t callback, void *user_data)
+{
+	bool reset = true;
+	struct counter_sam0_tc32_data *data = DEV_DATA(dev);
+	const struct counter_sam0_tc32_config *const cfg = DEV_CFG(dev);
+	TcCount32 *tc = cfg->regs;
+	int key = irq_lock();
+
+	if (data->ch.callback) {
+		goto err_busy;
+	}
+
+	if (callback) {
+		data->top_cb = callback;
+		data->top_user_data = user_data;
+		tc->INTENSET.reg = TC_INTFLAG_MC0;
+	} else {
+		tc->INTENCLR.reg = TC_INTFLAG_MC0;
+	}
+
+	tc->CC[0].reg = ticks;
+
+	if (reset) {
+		tc->CTRLBSET.reg = TC_CTRLBSET_CMD_RETRIGGER;
+	} else {
+		/*
+		 * Top trigger is on equality only, so manually reset it
+		 * if the counter has missed the new top.
+		 */
+		if (counter_sam0_tc32_read(dev) >= ticks) {
+			tc->CTRLBSET.reg = TC_CTRLBSET_CMD_RETRIGGER;
+		}
+	}
+	wait_synchronization(tc);
+
+	tc->INTFLAG.reg = TC_INTFLAG_MC0;
+	irq_unlock(key);
+	return 0;
+
+err_busy:
+	irq_unlock(key);
+	return -EBUSY;
+}
+
+static u32_t counter_sam0_tc32_get_pending_int(struct device *dev)
+{
+	const struct counter_sam0_tc32_config *const cfg = DEV_CFG(dev);
+	TcCount32 *tc = cfg->regs;
+
+	return tc->INTFLAG.reg & (TC_INTFLAG_MC0 | TC_INTFLAG_MC1);
+}
+
+static u32_t counter_sam0_tc32_get_top_value(struct device *dev)
+{
+	const struct counter_sam0_tc32_config *const cfg = DEV_CFG(dev);
+	TcCount32 *tc = cfg->regs;
+
+	/*
+	 * Unsync read is safe here because we're not using
+	 * capture mode, so things are only set from the CPU
+	 * end.
+	 */
+	return tc->CC[0].reg;
+}
+
+static u32_t counter_sam0_tc32_get_max_relative_alarm(struct device *dev)
+{
+	const struct counter_sam0_tc32_config *const cfg = DEV_CFG(dev);
```

Comment:
what about `return counter_sam0_tc32_get_top_value(dev) - 1;`

Reply:
Updated with all the requested changes.